src/python/pants/backend/docker/parser.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import re
+from dataclasses import InitVar, dataclass, field
+from typing import Generator, Pattern
+
+from dockerfile import Command, parse_string
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.engine.fs import DigestContents, PathGlobs
+from pants.engine.rules import Get, collect_rules, rule
+from pants.option.global_options import GlobalOptions
+
+
+@dataclass(frozen=True)
+class DockerfileParseRequest:
+    sources: DockerImageSources
+
+
+@dataclass(frozen=True)
+class DockerfileInfo:
+    putative_target_addresses: tuple[str, ...] = ()
+
+
+@dataclass
+class DockerfileParser:
+    dockerfile: InitVar[str]
+    commands: tuple[Command, ...] = field(init=False)
+
+    pex_target_regexp: str = r"""
+    (?# optional path, one level with dot-separated words)
+    (?:(?P<path>(?:\w[.0-9_-]?)+) /)?
+
+    (?# binary name, with .pex file extension)
+    (?P<name>(?:\w[.0-9_-]?)+) \.pex$
+    """
+
+    def __post_init__(self, dockerfile: str):
+        self.commands = parse_string(dockerfile)
+        self._compiled_pex_target_regexp = re.compile(self.pex_target_regexp, re.VERBOSE)
+
+    def get_all(self, command_name: str) -> Generator[Command, None, None]:
+        for command in self.commands:
+            if command.cmd.upper() == command_name:
+                yield command
+
+    @staticmethod
+    def translate_to_address(value: str, pex_target_regexp: Pattern) -> str | None:
```

Comment:
Passing the regex in is a bit weird. Probably better to encapsulate the compiled regex in an instance of this class, and not have this be a staticmethod after all.

Reply:
Yeah, I was trying a bunch of different stuff for that, and at one point had it as a classmethod, but came to the conclusion that that would not improve load times at all, so, regular method it is.